### PR TITLE
usb: device_next: add CDC ECM class implementation 

### PR DESCRIPTION
**Pull request by jfischer-no**
Add CDC Ethernet Control Model class implementation for the
new experimental USB device support based on the existing
implementation subsys/usb/device/class/netusb/function_ecm.c.

The implementation forms virtual ethernet connection between
USB host and USB device and requires two corresponding MAC
addresses, one for the virtual interface on the device side,
and other for the host which is described by a string descriptor.

CDC ECM implementation supports multiple instances which are
specified using DT. The number of instances is limited by the
number of endpoints on the controller, two to three should usually
be possible.

### CODE REVIEW

File: subsys/usb/device_next/class/usbd_cdc_ecm.c [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+
+	if (len > NET_ETH_MAX_FRAME_SIZE) {
+		LOG_WRN("Trying to send too large packet, drop");
+		return -ENOMEM;
+	}
+
+	if (!atomic_test_bit(&data->state, CDC_ECM_CLASS_ENABLED) ||
+	    !atomic_test_bit(&data->state, CDC_ECM_IFACE_UP)) {
+		LOG_INF("Configuration is not enabled or interface not ready");
+		return -EACCES;
```
Comment (tmon-nordic) [inline subsys/usb/device_next/class/usbd_cdc_ecm.c:438]: I think `ENODEV` is more appropriate here.

File: dts/bindings/ethernet/zephyr,cdc-ecm-ethernet.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+description: USB CDC ECM virtual ethernet controller
+
+compatible: "zephyr,cdc-ecm-ethernet"
+
+include: ethernet.yaml
+
+properties:
+  mac-address:
+    type: string
```
Comment (tmon-nordic) [inline dts/bindings/ethernet/zephyr,cdc-ecm-ethernet.yaml:12]: I think MAC address should be `required: true` property.

File: subsys/usb/device_next/class/usbd_cdc_ecm.c [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+	.nil_desc = {								\
+		.bLength = 0,							\
+		.bDescriptorType = 0,						\
+	},									\
+}
+
+#define USBD_CDC_ECM_DT_DEVICE_DEFINE(n)					\
+	CDC_ECM_DEFINE_DESCRIPTOR(n);						\
+	USBD_DESC_STRING_DEFINE(mac_desc_nd_##n,				\
+				DT_INST_PROP_OR(n, mac_address, "00005E005301"),\
```
Comment (tmon-nordic) [inline subsys/usb/device_next/class/usbd_cdc_ecm.c:696]: If MAC address is changed to require property this hardcoded fallback will no longer be necessary.

**Review by tmon-nordic — CHANGES_REQUESTED** [review]

File: dts/bindings/ethernet/zephyr,cdc-ecm-ethernet.yaml (hunk 2 above)
Comment (tmon-nordic) [reply dts/bindings/ethernet/zephyr,cdc-ecm-ethernet.yaml:12]: However, not sure how feasible it is to have mac address as a device tree property as these things are usually stored in some configuration memory.

File: subsys/usb/device_next/class/usbd_cdc_ecm.c [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+		net_if_carrier_off(data->iface);
+	}
+
+	if (data_iface == iface && alternate == 1) {
+		net_if_carrier_on(data->iface);
+		if (cdc_ecm_out_start(c_nd)) {
+			LOG_ERR("Failed to start OUT transfer");
+		}
+
+	}
```
Comment (GeorgeCGV) [inline subsys/usb/device_next/class/usbd_cdc_ecm.c:342]: Minor, could be simplified by factoring out :
```
if (data_iface != iface) {
	return;
}
```
then `else if` for the `alternate`.

File: subsys/usb/device_next/class/usbd_cdc_ecm.c [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+	struct cdc_ecm_eth_data *data = dev->data;
+
+	if (type == ETHERNET_CONFIG_TYPE_MAC_ADDRESS) {
+		memcpy(data->mac_addr, config->mac_address.addr,
+		       sizeof(data->mac_addr));
+
+		return 0;
+	}
+
+	return -ENOTSUP;
```
Comment (GeorgeCGV) [inline subsys/usb/device_next/class/usbd_cdc_ecm.c:490]: Could be simplified by inverting the condition:
```
if (type != ETHERNET_CONFIG_TYPE_MAC_ADDRESS) {
	return -ENOTSUP;
}

memcpy(data->mac_addr, config->mac_address.addr,
       sizeof(data->mac_addr));
return 0;
```

File: subsys/usb/device_next/class/usbd_cdc_ecm.c [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+	uint16_t ip_len;
+
+	if (len < NET_IPV6H_LEN + sizeof(struct net_eth_hdr)) {
+		/* Too short */
+		return 0;
+	}
+
+	switch (ntohs(hdr->type)) {
+	case NET_ETH_PTYPE_IP:
+	case NET_ETH_PTYPE_ARP:
```
Comment (GeorgeCGV) [inline subsys/usb/device_next/class/usbd_cdc_ecm.c:147]:
```suggestion
	case NET_ETH_PTYPE_IP:
		__fallthrough;
	case NET_ETH_PTYPE_ARP:
```

File: dts/bindings/ethernet/zephyr,cdc-ecm-ethernet.yaml (hunk 2 above)
Comment (jfischer-no) [reply dts/bindings/ethernet/zephyr,cdc-ecm-ethernet.yaml:12]: Maybe `remote-mac-address:` is a better property name.

File: subsys/usb/device_next/class/usbd_cdc_ecm.c (hunk 4 above)
Comment (GeorgeCGV) [reply subsys/usb/device_next/class/usbd_cdc_ecm.c:345]:
```suggestion
	if (data_iface != iface) {
		return;
	}

	if (alternate == 0) {
 		net_if_carrier_off(data->iface);
 	} else if (alternate == 1) {
		net_if_carrier_on(data->iface);
		if (cdc_ecm_out_start(c_nd)) {
			LOG_ERR("Failed to start OUT transfer");
		}
	}
```

Comment (jfischer-no) [reply subsys/usb/device_next/class/usbd_cdc_ecm.c:345]: It simplifies nothing.

**Review by tmon-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**